### PR TITLE
Use defer loading strategy in bundled themes

### PR DESCRIPTION
**Pull request by westonruter**
Trac ticket: https://core.trac.wordpress.org/ticket/59316

Current results for LCP-TTFB for 100 requests before & after as of 1648b8c:

Theme | Before | After | Diff (ms) | Percentage
-- | --: | --: | --: | --:
twentyeleven | 76.80 | 73.55 | -3.25 | -4.23%
twentytwelve | 78.35 | 77.50 | -0.85 | -1.08%
twentythirteen | 105.00 | 98.20 | -6.80 | -6.48%
twentyfourteen | 213.80 | 214.50 | +0.70 | +0.33%
twentyfifteen | 97.05 | 96.75 | -0.30 | -0.31%
twentysixteen | 99.00 | 101.05 | +2.05 | +2.07%
twentyseventeen | 72.85 | 75.20 | +2.35 | +3.23%
twentynineteen | 65.45 | 62.40 | -3.05 | -4.66%
twentytwenty | 77.50 | 77.70 | +0.20 | +0.26%
twentytwentyone | 55.35 | 53.90 | -1.45 | -2.62%
Average | -- | -- | -2.62 | -3.23%

These were computed using https://gist.github.com/westonruter/469c9de1ee699c97285e8bc8cc688991

The results are not very stable, however. Repeated runs seem to include a good deal of fluctuation.

# Drafted Commit Message

Bundled Themes: Use `defer` loading strategy for theme scripts.

* Add `defer` loading strategy for all frontend end-user theme scripts (excluding Customizer preview).
* Move scripts to the `head` which relate to the initial page viewport to ensure they start loading earlier and execute sooner while still not blocking rendering.
* Update Twenty Twenty's script loader (`TwentyTwenty_Script_Loader`) to support core's built-in script loading strategies (#12009), while also retaining backwards-compatibility for child themes that may set `async` and `defer` script data.
* Update the main script loading strategy in Twenty Twenty from `async` to `defer` for better performance on repeat page views, since when an `async` script is cached it will block rendering.

Props westonruter, flixos90, sabernhardt.
Fixes #59316.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-content/themes/twentytwenty/inc/template-tags.php [hunk 1]
```diff
@@ -659,7 +659,7 @@ function twentytwenty_nav_menu_social_icons( $item_output, $item, $depth, $args
 function twentytwenty_no_js_class() {
 
 	?>
-	<script>document.documentElement.className = document.documentElement.className.replace( 'no-js', 'js' );</script>
+	<script>document.documentElement.classList.remove("no-js");</script>
```
Comment (westonruter) [inline src/wp-content/themes/twentytwenty/inc/template-tags.php]: This is just a modernization change.

File: src/wp-content/themes/twentytwentyone/assets/js/primary-navigation.js [hunk 2]
```diff
@@ -198,7 +199,7 @@ function twentytwentyoneExpandSubMenu( el ) { // jshint ignore:line
 		} );
 	};
 
-	window.addEventListener( 'load', function() {
+	document.addEventListener( 'DOMContentLoaded', function() {
```
Comment (westonruter) [inline src/wp-content/themes/twentytwentyone/assets/js/primary-navigation.js]: The `load` event fires after all images load, meaning the menu could be uninitialized for a while after the DOM has loaded.

File: src/wp-content/themes/twentytwentyone/functions.php [hunk 3]
```diff
@@ -424,7 +424,7 @@ function twenty_twenty_one_scripts() {
 		array( 'in_footer' => true )
 	);
 
-	// Register the IE11 polyfill loader.
+	// Register the IE11 polyfill loader. This is no longer used but it is retained if any child theme or plugin uses them.
```
Comment (westonruter) [inline src/wp-content/themes/twentytwentyone/functions.php]: I removed `twenty-twenty-one-ie11-polyfills` from being a dependency for the other scripts.

File: src/wp-content/themes/twentytwentyone/functions.php [hunk 4]
```diff
@@ -447,23 +447,46 @@ function twenty_twenty_one_scripts() {
...
 	}
-
-	// Responsive embeds script.
-	wp_enqueue_script(
-		'twenty-twenty-one-responsive-embeds-script',
-		get_template_directory_uri() . '/assets/js/responsive-embeds.js',
-		array( 'twenty-twenty-one-ie11-polyfills' ),
-		wp_get_theme()->get( 'Version' ),
-		array( 'in_footer' => true )
-	);
```
Comment (westonruter) [inline src/wp-content/themes/twentytwentyone/functions.php]: I found this script was being enqueued unconditionally, even when it wasn't needed. So I introduced `twentytwentyone_enqueue_responsive_embeds_script_if_needed` below to only enqueue if the content contains an `iframe`.

File: src/wp-content/themes/twentytwentyone/functions.php [hunk 5]
```diff
@@ -635,7 +663,6 @@ function twentytwentyone_add_ie_class() {
 	</script>
 	<?php
 }
-add_action( 'wp_footer', 'twentytwentyone_add_ie_class' );
```
Comment (westonruter) [inline src/wp-content/themes/twentytwentyone/functions.php]: Since skip-link-focus-fix was deprecated, it seems sensible to do the same for this script.

File: src/wp-content/themes/twentytwentyone/inc/template-functions.php [hunk 6]
```diff
@@ -17,9 +17,6 @@
  */
 function twenty_twenty_one_body_classes( $classes ) {
 
-	// Helps detect if JS is enabled or not.
-	$classes[] = 'no-js';
```
Comment (westonruter) [inline src/wp-content/themes/twentytwentyone/inc/template-functions.php]: This has been moved to the `html` element for consistency with other themes. It is more performant on the document element, since if the class is modified on the `body` element, then this introduces the possibility of there being a re-layout due to a style change.

File: src/wp-content/themes/twentytwentyone/assets/js/primary-navigation.js [hunk 7]
```diff
@@ -115,7 +116,7 @@ function twentytwentyoneExpandSubMenu( el ) { // jshint ignore:line
 			mobileButton.onclick = function() {
 				wrapper.classList.toggle( id + '-navigation-open' );
 				wrapper.classList.toggle( 'lock-scrolling' );
-				twentytwentyoneToggleAriaExpanded( mobileButton );
+				twentytwentyoneToggleAriaExpanded( mobileButton, false );
```
Comment (westonruter) [inline src/wp-content/themes/twentytwentyone/assets/js/primary-navigation.js]: The function was getting static analysis complaints in PhpStorm since it was not provided. The function is using the falsiness of the `undefined` value.

File: src/wp-content/themes/twentytwenty/functions.php [hunk 8]
```diff
@@ -129,14 +129,6 @@ function twentytwenty_theme_support() {
...
 	// Add theme support for selective refresh for widgets.
 	add_theme_support( 'customize-selective-refresh-widgets' );
-
-	/*
-	 * Adds `async` and `defer` support for scripts registered or enqueued
-	 * by the theme.
-	 */
-	$loader = new TwentyTwenty_Script_Loader();
-	add_filter( 'script_loader_tag', array( $loader, 'filter_script_loader_tag' ), 10, 2 );
-
```
Comment (westonruter) [inline src/wp-content/themes/twentytwenty/functions.php]: In the case of child themes that register scripts with `defer` or `async` script data, should we automatically copy these to the `strategy` script data?

Here are the themes that add such script data: https://wpdirectory.net/search/01H9RXP8HXNAD30JZBJY4VRJZ2

However, there is only a single child theme of Twenty Twenty in the directory: https://wpdirectory.net/search/01H9RXTXYNFBGJD9PNY4QJ7CZV

And the `codoc` theme does does not register any of its own scripts.

File: src/wp-content/themes/twentynineteen/functions.php [hunk 9]
```diff
@@ -259,8 +259,26 @@ function twentynineteen_scripts() {
...
+				'strategy'  => 'defer',
+			)
+		);
+		wp_enqueue_script(
+			'twentynineteen-touch-navigation',
+			get_theme_file_uri( '/js/touch-keyboard-navigation.js' ),
+			array(),
+			'20230621',
+			array(
+				'in_footer' => false,
```
Comment (felixarntz) [inline src/wp-content/themes/twentynineteen/functions.php]: Is this worth being in the header? Thinking touch keyboard navigation isn't super critical to be immediately loaded?

File: src/wp-content/themes/twentynineteen/functions.php [hunk 10]
```diff
@@ -259,8 +259,26 @@ function twentynineteen_scripts() {
...
 	if ( has_nav_menu( 'menu-1' ) ) {
-		wp_enqueue_script( 'twentynineteen-priority-menu', get_theme_file_uri( '/js/priority-menu.js' ), array(), '20200129', array( 'in_footer' => true ) );
-		wp_enqueue_script( 'twentynineteen-touch-navigation', get_theme_file_uri( '/js/touch-keyboard-navigation.js' ), array(), '20230621', array( 'in_footer' => true ) );
+		wp_enqueue_script(
+			'twentynineteen-priority-menu',
+			get_theme_file_uri( '/js/priority-menu.js' ),
+			array(),
+			'20200129',
+			array(
+				'in_footer' => false,
```
Comment (felixarntz) [inline src/wp-content/themes/twentynineteen/functions.php]: You omitted the `// Because involves header.` here. Intentionally, or just an oversight?

File: src/wp-content/themes/twentytwenty/functions.php (hunk 8 above)
Comment (felixarntz) [reply src/wp-content/themes/twentytwenty/functions.php]: Yeah, I think that's a good idea. By removing this code here, we're not just deprecating but also breaking old usage by child themes, so making sure they still work is crucial. Migrating this data to the script loading strategy sounds great to me.

I am however wondering if deprecating overall is the way to go. What about pre-WP 6.3 sites? We would potentially degrade their performance if we just deprecate this and remove usage. I would prefer we make it an if-else thing based on the WP version and avoid the deprecation of that code (since it's still relevant pre-WP 6.3).

Maybe the cleanest solution would be:
* For any (child theme) script that adds `async` or `defer` script data, migrate to the `strategy` approach regardless of WP version.
* Update the `TwentyTwenty_Script_Loader` class (only relevant for pre-WP 6.3) to rely on `strategy`.

That way we can write the "new" code consistently, and it works across all versions.

File: src/wp-content/themes/twentytwenty/inc/template-tags.php (hunk 1 above)
Comment (felixarntz) [reply src/wp-content/themes/twentytwenty/inc/template-tags.php]: Why is that necessary? It changes the behavior as far as I can tell, the `.js` selector wouldn't work anymore. Some themes (including potentially this one?) have certain JS or CSS that targets such a selector, so I don't think removing it is worth the risk.

File: src/wp-content/themes/twentytwentyone/functions.php (hunk 4 above)
Comment (felixarntz) [reply src/wp-content/themes/twentytwentyone/functions.php]: This is unrelated to the main change here. While I think some of the JS load changes (e.g. using `DOMContentLoaded`) make sense in the scope of this ticket/PR, this change IMO goes too far and should be part of its own ticket, since it belongs into its own conversation.

I'm not opposed, but we shouldn't use this ticket as a general "let's fix bad performance things in bundled themes" :)

File: src/wp-content/themes/twentytwentyone/functions.php (hunk 5 above)
Comment (felixarntz) [reply src/wp-content/themes/twentytwentyone/functions.php]: See above, this is an unrelated change.

File: src/wp-content/themes/twentytwentyone/inc/template-functions.php (hunk 6 above)
Comment (felixarntz) [reply src/wp-content/themes/twentytwentyone/inc/template-functions.php]: See above, this is unrelated to this ticket/PR.

File: src/wp-content/themes/twentytwentyone/assets/js/responsive-embeds.js [hunk 11]
```diff
@@ -8,6 +8,7 @@
  * Add max-width & max-height to <iframe> elements, depending on their width & height props.
  *
  * @since Twenty Twenty-One 1.0
+ * @todo This may be obsolete now that aspect-ratio is a thing.
```
Comment (felixarntz) [inline src/wp-content/themes/twentytwentyone/assets/js/responsive-embeds.js]: See above, this comment is unrelated to this ticket.

**Review by felixarntz — CHANGES_REQUESTED** [review]
@westonruter Thank you for the PR, in terms of deferring it mostly looks good. My only feedback on that would be to find a cross-version compatible approach for the old Twenty Twenty hack.

Otherwise, you included several changes in here that have nothing to do with using the script loading strategy, so I think those should be removed and separate PRs / tickets opened for them.

File: src/wp-content/themes/twentytwenty/inc/template-tags.php (hunk 1 above)
Comment (sabernhardt) [reply src/wp-content/themes/twentytwenty/inc/template-tags.php]: The `.js` selector is necessary for Twenty Twenty.

https://github.com/WordPress/wordpress-develop/blob/1118e37f48ff184c40bc4e593ae13b1fbe4b8001/src/wp-content/themes/twentytwenty/style.css#L1056-L1062

With other themes that replace `no-js` with `js`, even if the theme itself does not use the class, someone might use it in a child theme (as I have done with Twenty Fifteen and Twenty Sixteen).

Comment (westonruter) [reply src/wp-content/themes/twentytwenty/inc/template-tags.php]: Oh yes, I neglected to add the `js`. Nevertheless, I'll just revert this change.

File: src/wp-content/themes/twentynineteen/functions.php (hunk 9 above)
Comment (westonruter) [reply src/wp-content/themes/twentynineteen/functions.php]: Good point.

File: src/wp-content/themes/twentynineteen/functions.php (hunk 10 above)
Comment (westonruter) [reply src/wp-content/themes/twentynineteen/functions.php]: Yes, an oversight.

File: src/wp-content/themes/twentytwenty/functions.php (hunk 8 above)
Comment (westonruter) [reply src/wp-content/themes/twentytwenty/functions.php]: OK, I've implemented this in 1648b8c3077d2e7b15db4ffae33d195198a14cb5.

File: src/wp-content/themes/twentytwenty/classes/class-twentytwenty-script-loader.php [hunk 12]
```diff
@@ -33,10 +55,17 @@ class TwentyTwenty_Script_Loader {
...
-			foreach ( array( 'async', 'defer' ) as $attr ) {
-				if ( ! wp_scripts()->get_data( $handle, $attr ) ) {
-					continue;
-				}
+			$strategies = array(
+				'async' => (bool) wp_scripts()->get_data( $handle, 'async' ),
+				'defer' => (bool) wp_scripts()->get_data( $handle, 'defer' ),
+			);
+			$strategy   = wp_scripts()->get_data( $handle, 'strategy' );
+			if ( in_array( $strategy, array( 'async', 'defer' ), true ) ) {
```
Comment (felixarntz) [inline src/wp-content/themes/twentytwenty/classes/class-twentytwenty-script-loader.php]: Nit-pick, and 0.0000001% faster :)
```suggestion
			if ( $strategy && isset( $strategies[ $strategy ] ) ) {
```

File: src/wp-content/themes/twentytwenty/functions.php [hunk 13]
```diff
@@ -211,8 +215,16 @@ function twentytwenty_register_scripts() {
...
 
-	wp_enqueue_script( 'twentytwenty-js', get_template_directory_uri() . '/assets/js/index.js', array(), $theme_version );
-	wp_script_add_data( 'twentytwenty-js', 'async', true );
+	wp_enqueue_script(
+		'twentytwenty-js',
+		get_template_directory_uri() . '/assets/js/index.js',
+		array(),
+		$theme_version,
+		array(
+			'in_footer' => false,
```
Comment (felixarntz) [inline src/wp-content/themes/twentytwenty/functions.php]:
```suggestion
			'in_footer' => false, // Because involves header.
```

File: src/wp-content/themes/twentytwentyone/functions.php [hunk 14]
```diff
@@ -449,7 +449,10 @@ function twenty_twenty_one_scripts() {
 			get_template_directory_uri() . '/assets/js/primary-navigation.js',
 			array( 'twenty-twenty-one-ie11-polyfills' ),
 			wp_get_theme()->get( 'Version' ),
-			array( 'in_footer' => true )
+			array(
+				'in_footer' => false,
```
Comment (felixarntz) [inline src/wp-content/themes/twentytwentyone/functions.php]:
```suggestion
				'in_footer' => false, // Because involves header.
```

File: src/wp-content/themes/twentytwenty/functions.php [hunk 15]
```diff
@@ -211,8 +215,16 @@ function twentytwenty_register_scripts() {
...
+			'strategy'  => 'defer',
```
Comment (felixarntz) [inline src/wp-content/themes/twentytwenty/functions.php:225]: Just double-checking: This was `async`, but is now `defer`. Is there a chance for any unexpected behavior due to the loading order? If so, have you tested that?

**Review by felixarntz — APPROVED** [review]
Thanks @westonruter. Basically LGTM, just a few nit-picks.

File: src/wp-content/themes/twentytwenty/classes/class-twentytwenty-script-loader.php (hunk 12 above)
Comment (westonruter) [reply src/wp-content/themes/twentytwenty/classes/class-twentytwenty-script-loader.php]: I like it. Reusing a data structure we already have.

File: src/wp-content/themes/twentytwenty/functions.php (hunk 15 above)
Comment (westonruter) [reply src/wp-content/themes/twentytwenty/functions.php:225]: Good to double-check. Since the loading order was not guaranteed before with `async` whereas now it is always right before `DOMContentLoaded`, if anything it will be more consistent (and it won't block rendering if the script is cached). The script's [initialization logic](https://github.com/WordPress/wordpress-develop/blob/827a560adef5e5f1d9fea1788b58aab9c4752a10/src/wp-content/themes/twentytwenty/assets/js/index.js#L653-L663) already accounts for whether the script is loaded before or after `DOMContentLoaded` due to the previous `async` loading strategy, so it should be all good on that front. In any case, there are no dependencies nor dependents of this script, so there won't be any loading order issue regarding scripts. Loading a page in the browser, there are no errors and the functionality works as expected.

Comment (felixarntz) [reply src/wp-content/themes/twentytwenty/functions.php:225]: Thanks for clarifying, makes sense!

**Comment by westonruter** [comment]
Committed in r56556.

File: src/wp-content/themes/twentyeleven/showcase.php [hunk 16]
```diff
@@ -16,7 +16,16 @@
...
-wp_enqueue_script( 'twentyeleven-showcase', get_template_directory_uri() . '/js/showcase.js', array( 'jquery' ), '20211130' );
+wp_enqueue_script(
+	'twentyeleven-showcase',
+	get_template_directory_uri() . '/js/showcase.js',
+	array( 'jquery' ),
+	'20211130',
+	array(
+		'in_footer' => false, // Because involves header.
+		'strategy'  => 'defer',
+	)
```
Comment (westonruter) [inline src/wp-content/themes/twentyeleven/showcase.php:27]: Oh, this actually was originally added to the `head`, so in WP<6.3 this will actually get printed in the footer since the `$args` array is truthy for `$in_footer`. Nevertheless, the function's contents are wrapped in `jQuery.ready()` so it makes no difference when it loads.
